doc/developer/design/20240625_subsource_deprecation_blue_green_sources.md

```diff
@@ -0,0 +1,317 @@
+
+If they'd like to ingest data from both an upstream Postgres/MySQL database
+and a Kafka topic, they also need to learn the fragmented user interface
+for dealing with data ingested from upstream data sources. Our existing
+model of a *source* does not provide a consistent interface among source types.
+
+Single-output Kafka sources put their data under the source's primary relation,
+whereas multi-output Postgres/MySQL sources do not allow querying the source's primary
+relation and instead contain *subsources*, an entirely separate concept.
+
+Our existing *source* models also make it difficult to deal with upstream schema changes,
+such that users need to drop and recreate a *source* or *subsource* and any dependent objects
+to begin ingesting data with a new schema. This can cause long-periods of downtime if the
+new source needs to reingest the entire upstream history. This can also cause them to
+lose their _reclocking_ information which makes it impossible to implement a correct
+end-to-end *source* -> *sink* pipeline.
+
+This change aims to simplify the user experience and create a more unified
+model for managing data ingested from an upstream source, while providing
+more flexibility in handling upstream schema changes.
+
+
+## Success Criteria
+
+- The mental overhead & learning time for users to ingest data from an upstream system
+  is decreased.
+- All source types use a consistent interface to retrieve progress information and
+  ingested data. The primary *source* object represents the same concept across
+  all source types.
+- The same upstream table/topic can be ingested into multiple collections in Materialize,
+  each with a potentially different projection of the upstream schema. This enables
+  a user to do blue-green schema-change workflows without unavailability.
+- *sources* can maintain their progress collection while the set of collections being
+  ingested are modified, allowing an end-to-end *source* -> *sink* schema-change workflow
+  that does not emit redundant data (reusable reclocking).
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+- Accomplishing 'subsource demuxing' using the concept of *subsources* in-use today.
+  The goals of https://github.com/MaterializeInc/materialize/issues/24843 will be
+  accomplished during the process of replacing *subsources* with *tables* (see
+  solution proposal below).
+
+- Dealing with Webhook sources. These are fundamentally different than the other
+  storage-owned source types whose operators implement the `SourceRender` trait,
+  and they do not contain a 'progress' collection. While they may also be
+  converted to use 'tables' in the future, that work is not in-scope for this proposal.
```

Comment:
I do agree that we could convert webhook sources to this same 'table' model, but it's not necessary for this proposal and is probably something we should revisit when moving webhook sources to `clusterd`

Reply:
I agree, not necessary for this proposal, but it would be nice to have a consistent model from the Console's perspective. For example, we're building a database object explorer where tables have a different UI (shows columns) than sources:

<img width="500" alt="Screenshot 2024-07-02 at 5 35 00 PM" src="https://github.com/MaterializeInc/materialize/assets/29664023/c682cd58-e207-4924-bbcc-53b9646a23a3">

Right now we'd have to special case webhook sources to show columns due to this inconsistency

Reply:
Yeah, I wanted to check in on it because it will stick out like a sore thumb in multiple places, including the documentation and the Console. cc @chaas for aligment here.